### PR TITLE
Adds support for Content Templates (aka Blueprints)

### PR DESCRIPTION
**Pull request by leekelleher**
Reworks how the Content Type picker overlay works.

Targets the overlay to the "right" panel and enables Content Type selection to follow the same pattern as creating regular content nodes.

When a Content Type is selected, any Content Templates (Blueprints) will be displayed and can be selected.

Underlying updates have been made to the InnerContent resource & service scripts - these may be breaking changes, though I have only tested using Stacked Content, which (at present) has no side effects.

### CODE REVIEW

**Comment by leekelleher** [comment]
@mattbrailsford I'm sure you'll have a lot of questions - and Angular advice 😉

File: src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs [hunk 1]
```diff
@@ -75,10 +76,17 @@ public IEnumerable<object> GetContentTypesByAlias([ModelBinder] string[] aliases
...
-        public ContentItemDisplay GetContentTypeScaffoldByGuid(Guid guid)
+        public ContentItemDisplay GetContentTypeScaffoldByGuid(Guid contentTypeGuid, int blueprintId = 0)
         {
-            var contentType = Services.ContentTypeService.GetContentType(guid);
-            return new ContentController().GetEmpty(contentType.Alias, -20);
+            var controller = new ContentController();
+
+            if (blueprintId > 0)
+            {
+                return controller.GetEmpty(blueprintId, -20);
```
Comment (mattbrailsford) [inline src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs]: If supplying the blueprintId means the guid isn't needed, I'm wondering if this should mean 2 seperate methods should be present instead GetByGuid and GetByBlueprintId?

File: src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js [hunk 2]
```diff
@@ -416,9 +473,9 @@ angular.module("umbraco").factory('innerContentService', [
             return icResources.getContentTypeIconsByGuid(guids);
         }
 
-        self.createEditorModel = function (contentType, dbModel) {
+        self.createEditorModel = function (contentType, blueprintId, dbModel) {
```
Comment (mattbrailsford) [inline src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js]: Any changes made to the publish api of innerContentService should be classed as a breaking change

File: src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js [hunk 3]
```diff
@@ -111,44 +167,46 @@ angular.module('umbraco.directives').directive('innerContentOverlay', [
             scope.config.editorModels = scope.config.editorModels || {};
             scope.currentItem = null;
 
-            var getContentType = function (guid) {
+            var getContentType = function (key) {
```
Comment (mattbrailsford) [inline src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js]: Why is this changed to key? Is this now a cachekey of guid + blueprint id? If so, does that mean the inner equality check wouldn't be right?

File: src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs [hunk 4]
```diff
@@ -32,18 +32,19 @@ public IEnumerable<object> GetAllContentTypes()
...
-                });
+            var contentTypes = Services.ContentTypeService.GetAllContentTypes(guids).OrderBy(x => x.SortOrder).ToList();
+            var blueprints = Services.ContentService.GetBlueprintsForContentTypes(contentTypes.Select(x => x.Id).ToArray()).ToArray();
+
+            // NOTE: Using an anonymous class, as the `ContentTypeBasic` type is heavier than what we need (for our requirements)
+            return contentTypes.Select(ct => new
+            {
+                name = ct.Name, // TODO: localize the name (in case of dictionary items), e.g. `localizedTextService.UmbracoDictionaryTranslate`
+                description = ct.Description, // TODO: localize the description (in case of dictionary items), e.g. `localizedTextService.UmbracoDictionaryTranslate`
+                key = ct.Key,
```
Comment (mattbrailsford) [inline src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs]: Change from "guid" to "key", need to check if this is every exposed outside of inner content as if so, this is a breaking change.

File: src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs (hunk 1 above)
Comment (leekelleher) [reply src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs]: I'd had a thought that we could try to use a single `Guid` for both ContentType and Blueprint - and the controller method can figure out which is which?

But I'm also happy for additional method to specify ContentType or Blueprint.

File: src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js (hunk 2 above)
Comment (leekelleher) [reply src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js]: I was thinking of v2'ing this. But yes, if there's a way to not have a breaking-change, that'd be ace - I'll re-review the code.

File: src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs (hunk 1 above)
Comment (mattbrailsford) [reply src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs]: I think the seperate methods would be more explicit so easier to know what is going on

File: src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js (hunk 3 above)
Comment (leekelleher) [reply src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js]: Umbraco keeps referring to the Guid as Key - was trying to be consistent.

With the cachekey - I'd hit an issue where if you first selected a ContentType, then selected a Blueprint (or vice-versa), the first one would be cached. So they needed different cachekeys.

There may be a better way? (Hence requesting a review 😉)

File: src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs (hunk 4 above)
Comment (leekelleher) [reply src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs]: We can revert back to `guid` if you prefer? But yeah a breaking-change.
It's not impacted Stacked Content - which has been our only known implementation.

File: src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js (hunk 3 above)
Comment (mattbrailsford) [reply src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js]: I don't mind going with consistency then, so should be ok to update these methods as you have, but I might be tempted to expose a guid and key from the api endpoint for backwards compatibilities sake?

File: src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs (hunk 4 above)
Comment (mattbrailsford) [reply src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs]: Hmm, we could really do with knowing if anyone else is using inner content yet. My gut says no, in which case, it probably doesn't matter if we change it.

Comment (leekelleher) [reply src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs]: Well, there's nothing "official" (e.g. published on NuGet) 😃 

- http://nugetmusthaves.com/Dependency/Our.Umbraco.InnerContent
- http://nugetmusthaves.com/Dependency/Our.Umbraco.InnerContent.Core

Still, I'd like to "do the right thing" in terms of handling API changes.

Comment (mattbrailsford) [reply src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs]: If that's the case, then maybe have a guid and a key?

File: src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs [hunk 5]
```diff
@@ -32,18 +32,19 @@ public IEnumerable<object> GetAllContentTypes()
...
+                icon = string.IsNullOrWhiteSpace(ct.Icon) || ct.Icon == ".sprTreeFolder" ? "icon-document" : ct.Icon,
+                blueprints = blueprints.Where(bp => bp.ContentTypeId == ct.Id).ToDictionary(bp => bp.Id, bp => bp.Name)
+                // TODO: tabs = ct.CompositionPropertyGroups.Select(y => y.Name).Distinct()
```
Comment (leekelleher) [inline src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs]: @mattbrailsford On this line, I can't find anywhere that references the `tabs` field (from the call to `GetContentTypesByGuid`).  It's probably a breaking-change to remove it, but what are your thoughts?

Comment (mattbrailsford) [reply src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs]: hmm, are tabs fetched later on for the dialog window then?

File: src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js [hunk 6]
```diff
@@ -350,8 +407,8 @@ angular.module("umbraco").factory('innerContentService', [
 
         var self = {};
 
-        var getScaffold = function (contentType) {
-            return icResources.getContentTypeScaffoldByGuid(contentType.icContentTypeGuid).then(function (scaffold) {
+        var getScaffold = function (contentType, blueprintId) {
+            return icResources.getContentTypeScaffoldByGuid(contentType.icContentTypeGuid, blueprintId).then(function (scaffold) {
```
Comment (leekelleher) [inline src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js]: @mattbrailsford If we're going to have 2 controller methods (one for Guid, the other for BlueprintId), then how should I handle this here? Given the return method is from `getContentTypeScaffoldByGuid`'s promise?

Would it be to separate out the contents of the `then(function)`?

File: src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs (hunk 5 above)
Comment (leekelleher) [reply src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs]: Yeah, they come back with the scaffold object.

File: src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js (hunk 6 above)
Comment (mattbrailsford) [reply src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js]: yea, put the function in a local variable beforehand, then use it in either `then` call depending on which endpoing you need to call

File: src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs (hunk 5 above)
Comment (mattbrailsford) [reply src/Our.Umbraco.InnerContent/Web/Controllers/InnerContentApiController.cs]: Ahh, ok, then yea, looks like they aren't needed at this point then. Technically another breaking change though yea, but again, I doubt anyone is using it :)

File: src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js (hunk 2 above)
Comment (leekelleher) [reply src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js]: @mattbrailsford I've updated the PR, but not sure what to do about this one.

Comment (mattbrailsford) [reply src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js]: If we wanna maintain backwards compat, I'd say move it to the end of the method signature, or, run some code to see if it's a valid blueprintId or a dbModel in position two, if it's a dbModel, then set the dbModel var value if you get what I mean?

Comment (leekelleher) [reply src/Our.Umbraco.InnerContent/Web/UI/App_Plugins/InnerContent/js/innercontent.controllers.js]: PR updated - I've moved it to the end of the method signature.  I think checking the object-type of the 2nd argument feels a bit urgh.

**Comment by leekelleher** [comment]
Ace! Thanks for reviewing @mattbrailsford 🎉
